library/common/extensions/filters/http/platform_bridge/filter.cc

```diff
@@ -159,25 +179,64 @@ PlatformBridgeFilter::onTrailers(Http::HeaderMap& trailers,
 
   envoy_headers in_trailers = Http::Utility::toBridgeHeaders(trailers);
   envoy_filter_trailers_status result = on_trailers(in_trailers, platform_filter_.instance_context);
-  Http::FilterTrailersStatus status = static_cast<Http::FilterTrailersStatus>(result.status);
-  // TODO(goaway): Current platform implementations expose immutable trailers, thus any modification
-  // necessitates a full copy. Add 'modified' bit to determine when we can elide the copy. See also
-  // https://github.com/lyft/envoy-mobile/issues/949 for potential future optimization.
-  trailers.clear();
-  for (envoy_header_size_t i = 0; i < result.trailers.length; i++) {
-    trailers.addCopy(
-        Http::LowerCaseString(Http::Utility::convertToString(result.trailers.headers[i].key)),
-        Http::Utility::convertToString(result.trailers.headers[i].value));
-  }
-  // The C envoy_trailers struct can be released now because the trailers have been copied.
-  release_envoy_headers(result.trailers);
-  return status;
+
+  switch (result.status) {
+  case kEnvoyFilterTrailersStatusContinue:
+    if (iteration_state_ == IterationState::Stopped) {
+      PANIC("invalid filter state: ResumeIteration  may only be used when filter iteration is "
+            "stopped");
+    }
+    PlatformBridgeFilter::replaceHeaders(trailers, result.trailers);
+    return Http::FilterTrailersStatus::Continue;
+
+  case kEnvoyFilterTrailersStatusStopIteration:
+    iteration_state_ = IterationState::Stopped;
+    return Http::FilterTrailersStatus::StopIteration;
+
+  case kEnvoyFilterTrailersStatusResumeIteration:
+    if (iteration_state_ != IterationState::Stopped) {
+      PANIC("invalid filter state: ResumeIteration  may only be used when filter iteration is "
+            "stopped");
+    }
+    if (result.extra_headers) {
+      PlatformBridgeFilter::replaceHeaders(**pending_headers, *result.extra_headers);
+      *pending_headers = nullptr;
+      free(result.extra_headers);
+    }
+    if (result.extra_data) {
```

Comment:
`extra_data` is also a bit deceiving since it's a replacement set of data

Reply:
I'll rename it to `pending_data`.